---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## R Markdown

This is an R Markdown document. Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents. For more details on using R Markdown see <http://rmarkdown.rstudio.com>.

When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:

# load libraries
```{r}
library(openxlsx)
library(ggplot2)
library(tidyverse)
library(reprex)
library(matrixStats)
library(XML)
library(ggrepel)
library(DESeq2)
library(ComplexHeatmap)
library(dplyr)
library(readxl)
library(tximport)
library(edgeR)
library(openxlsx)
library(readxl)
library(readr)
library(apeglm)

source('X:/charliew/RNAseq-Class/R_code/data_formatting_tools.R')
source('X:/charliew/RNAseq-Class/R_code/analysis_tools.R')
source('X:/charliew/RNAseq-Class/R_code/plots.R')

options(scipen = 999)
```

# location and initialize file locations
```{r}
setwd("X:/charliew/RNAseq-Class/")

metadata_dir <- "X:/charliew/RNAseq-Class/metadata/" # diractory to store infosite metadata and QC figures
salmon_dir <- 'X:/charliew/RNAseq-Class/salmon/' # directory to store salmon script alignment files from salmon
deseq_dir <- 'X:/charliew/RNAseq-Class/deseq2/' # directory to store deseq2 results
figure_dir <- 'X:/charliew/RNAseq-Class/figures/'
data_dir <- 'X:/charliew/RNAseq-Class/data/'
gsea_dir <- 'X:/charliew/RNAseq-Class/gsea/'

alignment_target <- "/net/ostrom/data/bcc/charliew/RNAseq-Class/hg38_ens105_1.6/hg38_ens105/" # salmon alignment target

# tx2genefile <- "X:/dpradhan/pipelinedev/mm10_ens100_geneID_to_Type_MGI.xlsx"
orthologyfile <- ""
tx2genefile <- 'X:/charliew/RNAseq-Class/documents/hg38ens105_txID_geneID.xlsx' #'X:/charliew/annotationFiles/mm10_ens101_txID_to_geneID.xlsx' #"mm10_ens100_geneID_to_Type_MGI.xlsx"
id2symfile <- 'X:/charliew/RNAseq-Class/documents/hg38ens105_geneID_Sym_Type.xlsx'
infosite_loc <- '' # this should ideally be a /net/ostrom/... etc full path to the infosite qc file
```

# read in metadata
```{r}
setwd(metadata_dir)
metadata <- read_excel('metadata.xlsx') %>% as.data.frame()
```

# process tximport output
```{r}
setwd(salmon_dir)

salmon_data <- salmon_dat #read.table("summarized.txt", header = TRUE)

id2type2sym <- read.xlsx(id2symfile)
# HuMo_Orthology <- read.xlsx(orthologyfile)

#a subset of Mouse ensembl gene IDs have more than one assigned gene symbol
#these are ambiguous annotations and may potentially confound downstream analyses
#remove tehm from the annotation file
# mm10_id2type2sym <- mm10_id2type2sym %>% filter(!Gene.stable.ID %in% c("ENSMUSG00000094121","ENSMUSG00000096271", "ENSMUSG00000115016"))
#attach gene type and symbol data to the expression file
salmon_dat_named <- merge(salmon_data, id2type2sym, by.x="geneid", by.y = "Gene.stable.ID", all.x = FALSE)
#calculate row average and variance and join to the data
salmon_dat_named <- salmon_dat_named %>% mutate(l2tpmAv = rowMeans(dplyr::select(salmon_dat_named,contains(".l2tpm"))))
salmon_dat_named <- salmon_dat_named %>% mutate(l2tpmVar =rowVars(as.matrix(dplyr::select(salmon_dat_named,contains(".l2tpm")))))
#calculate overall average and variance and assign Yes/No Max Per Gene (MPG) flag
salmon_dat_named <- calculate_MPG(salmon_dat_named, "HGNC.symbol", avg_col = 'l2tpmAv', var_col = 'l2tpmVar', count_cols = '.l2tpm', mpg_col = 'MPG')

#annotate expression data with Human Ortholog
# salmon_dat_named <- add_orthology(salmon_dat_named, HuMo_Orthology,
#                                   match_column = 'MGI.symbol', orthology_name = 'HuSym',
#                                   ref_col = 2, orth_col = 1)
#add huMPG column
# salmon_dat_named <- calculate_MPG(salmon_dat_named, "HuSym", count_cols = '.l2tpm', mpg_col = 'huMPG')

salmon_dat_named <- salmon_dat_named %>% mutate(LowAvg = l2tpmAv <= 1)
salmon_dat_named <- salmon_dat_named %>% mutate(LowVar = l2tpmVar <= 0.1)
```

# add sample names and groups to metadata for DESeq
```{r}
setwd(deseq_dir)

sample_names <- colnames(salmon_dat_named) %>% grep('intCt', x=., value = TRUE) %>% gsub('\\.intCt','', x = .)
metadata$name <- sample_names
# metadata$group <- strsplit(metadata$name %>% as.character(), split = '\\.') %>% sapply(., "[[", 1) %>% as.factor

```

# heatmap
```{r}
setwd(figure_dir)

heatmap_dat <- salmon_dat_named[which(salmon_dat_named$MPG == 'Yes' & salmon_dat_named$LowAvg == FALSE & salmon_dat_named$LowVar == FALSE),
                                grep('intCt$', colnames(salmon_dat_named))] %>% as.matrix()
# %>% rename_at(colnames(.), ~ metadata$`Sample name`[order(metadata$Sample.ID)])
pdf(file = "heatmap.pdf",   # The directory you want to save the file in
    width = 20, # The width of the plot in inches
    height = 30)
Heatmap(heatmap_dat, column_names_side = 'top', row_names_gp = gpar(fontsize = 6), column_names_gp = gpar(fontsize = 9), 
        name = 'l2cpm',
        rect_gp = gpar(col = "white", lwd = 0.01),
        column_title = '')
dev.off()
```

# pca
```{r}
setwd(figure_dir)

pca <- salmon_dat_named[which(salmon_dat_named$MPG == 'Yes' & salmon_dat_named$LowAvg == FALSE & salmon_dat_named$LowVar == FALSE),] %>% 
  full_pca(., gene_col = 'HGNC.symbol')
# var <- samples$treatment
rownames(pca$x) <- gsub(pattern = '.l2tpm', replacement = '', x = rownames(pca$x))
# PC <- merge(pca$x, metadata, by=0) #data.frame(pca$x, var=var)
PC <- cbind(pca$x, metadata[order(metadata$Sample),])
ggplot(PC,aes(x=PC1,y=PC2, color = paste(Condition)))+
  geom_point(size=3)+
  theme_classic()+ggtitle('PCA')
ggsave(filename = 'PCA_1_2.pdf', width = 15, height = 10)
```

# Individual Gene check
```{r}
gene <- 'Selenop'

intCt <- salmon_dat_named[which(salmon_dat_named$MGI.symbol == gene) ,grepl('\\.intCt', colnames(salmon_dat_named))]
l2tpm <- salmon_dat_named[which(salmon_dat_named$MGI.symbol == gene) ,grepl('\\.l2tpm', colnames(salmon_dat_named))]

print(paste(dim(intCt)[1], dim(l2tpm)[1]))

df <- data.frame(intCt = unlist(intCt), l2tpm = unlist(l2tpm), Name = names(intCt) %>% strsplit(., split = '\\.') %>% lapply(., '[[', 1) %>% unlist())

data <- data.frame(Name = metadata$Sample.name,
                      Diet = metadata$Diet, 
                      `Cell type` = metadata$Cell.type, Genotype = metadata$Genotype,
                      Age = metadata$Age, Replicate =metadata$Replicate,
                      Group = metadata$Group)
data <- merge(data, df, by = 'Name')

ggplot(data, mapping = aes(x = Group, y = intCt, color = Genotype)) + geom_boxplot() +
  theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggtitle(gene)

ggplot(data, mapping = aes(x = Group, y = l2tpm, color = Genotype)) + geom_boxplot() +
  theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggtitle(gene)

ggplot(data, mapping = aes(x = Name, y = l2tpm, color = Group)) + geom_line() + geom_point() +
  theme_classic() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
  ggtitle(gene)

```

# setup DESeq object and write metadata
```{r}
setwd(metadata_dir)

data <- salmon_dat_named
data <- data[-which(duplicated(data$geneid)),]

cnts <- data[,grepl('intCt', colnames(data))]
colnames(cnts) <- gsub('\\.intCt','', colnames(cnts))
colnames(cnts) <- colnames(cnts) %>% gsub(pattern = '150930AmoA_D15-', replacement = 'd',.) %>% gsub(pattern = '.gz', replacement = '',.)
rownames(cnts) <- data$geneid

l2tpm <- data[,grepl('\\.l2tpm', colnames(data))]
colnames(l2tpm) <- gsub('\\.l2tpm','', colnames(l2tpm))
colnames(l2tpm) <- colnames(l2tpm) %>% gsub(pattern = '150930AmoA_D15-', replacement = 'd',.) %>% gsub(pattern = '.gz', replacement = '',.)
rownames(l2tpm) <- data$geneid

samples <- data.frame(row.names = metadata$Sample, 
                      Condition = metadata$Condition %>% as.factor)

```

# heatmap for plotting gene sets
```{r}
# setwd(figure_dir)
# salmon_dat_named <- read_excel(paste0(salmon_dir, 'salmon_data.xlsx')) %>% as.data.frame()

gene_set <- c('Ide', 'Lgr5', 'Als2')

heatmap_dat <- salmon_dat_named[which(salmon_dat_named$MPG == 'Yes' & salmon_dat_named$MGI.symbol %in% gene_set),]
rownames(heatmap_dat) <- heatmap_dat$MGI.symbol
heatmap_dat <- heatmap_dat[,grep('l2tpm$', colnames(salmon_dat_named))] %>% as.matrix()
colnames(heatmap_dat) <- colnames(heatmap_dat) %>% strsplit(., split = '\\.') %>% lapply(., '[[', 1) %>% unlist()
# pdf(file = "Feeding_Heatmap.pdf",   # The directory you want to save the file in
#     width = 10, # The width of the plot in inches
#     height = 20)
Heatmap(heatmap_dat, column_names_side = 'top', row_names_gp = gpar(fontsize = 6), column_names_gp = gpar(fontsize = 9), 
        name = 'l2cpm',
        rect_gp = gpar(col = "white", lwd = 0.15),
        column_title = '')
# dev.off()
```

# set DESeq comparisons: Rev v G
```{r}
setwd(deseq_dir)

exp_cnts <- cnts
exp_l2tpm <- l2tpm
exp_samples <- samples#[colnames(exp_cnts),]
exp_samples$Condition <- relevel(exp_samples$Condition, "CS")

design <- ~Condition

date <- gsub(pattern = '-', replacement = '_', x = Sys.Date())
dir.create(file.path(deseq_dir, date))

setwd(paste0(deseq_dir, date))
write_metadata_record(design = design, metadata_samples = exp_samples,
                      annotation_file = tx2genefile, orthology_file = orthologyfile, infosite_file = infosite_loc,
                      sample_file_name = 'samples.csv')

all(colnames(cnts) == rownames(samples)) %>% print()

# dds <- DESeqDataSetFromMatrix(exp_cnts, colData=exp_samples, design)
dds <- DESeqDataSetFromTximport(txi, colData = exp_samples, design)
dds <- DESeq(dds, minReplicatesForReplace = 3)
print(exp_samples)
resultsNames(dds) %>% print()

```

# run DESeq comparisons
```{r}
setwd(paste0(deseq_dir, date))

comparisons <- c(2)

for (i in comparisons){
  res_ape <- lfcShrink(dds, coef=i, type="apeglm")
  
  apeData <- as.data.frame(res_ape)
  compar_name <- resultsNames(dds)[i]
  write(compar_name, file = 'run_record.txt', append = TRUE)
  stat_col <- paste('WaldStatistic', compar_name, sep = '_')
  stat <- dds@rowRanges@elementMetadata[,stat_col]
  
  res <- as.data.frame(res_ape)
  res$stat <- stat
  res <- cbind(geneid = rownames(res), res)
  write.table(as.data.frame(res), file=paste(compar_name, '_results.tsv', sep = ''), quote=FALSE, sep="\t", row.names = FALSE)
}

```

# process DESeq2 output (updated)
```{r}
setwd(paste0(deseq_dir, '2022_01_23/'))
num_base_cols <- dim(salmon_dat_named)[2]
result_files <- list.files(recursive = TRUE) %>% grep(pattern = '_results.tsv', ., value = TRUE)
deseq_data <- aggregate_deseq_results(result_files = result_files, count_data = salmon_dat_named)
deseq_data <- arrange_deseq_results(deseq_data, base_cols = 1:num_base_cols) #merge(count_data, compar, by = 'geneid')
write.csv(deseq_data, file = '150930Amo.csv', row.names = FALSE)

```

# read in deseq data
```{r}
setwd(paste0(deseq_dir, '2021_11_08/'))
deseq_data <- read.csv('210909Yil_CpsKO.csv')
```


# plot volcano 
```{r}
setwd(figure_dir)
options(ggrepel.max.overlaps = Inf)

data <- deseq_data[which(deseq_data$MPG == 'Yes' & !deseq_data$LowAvg & !deseq_data$LowVar & deseq_data$Gene.type == "protein_coding"),]

comparisons <- c('Condition_G_vs_Rev')

for (i in comparisons){
  volcano_plot(data, comparison = i, 
             adp_identifier = 'padj', lfc_identifier = 'log2FoldChange', 
             gene_column = "HGNC.symbol")
  ggsave(filename = paste0(i, '.pdf'), width = 8, height = 8)
}

# lfc_compar_plot(data, comparisons = c('Group_Aged_HPD_GFP.high_vs_Aged_Control_GFP.high', 'Group_Aged_HPD_GFP.low_vs_Aged_Control_GFP.low'), 
#              adp_identifier = 'padj', lfc_identifier = 'log2FoldChange', 
#              gene_column = "HGNC.symbol")
# ggsave(filename = paste0('COndition', '.pdf'), width = 8, height = 8)


```

# prep IPA and GSEA files
```{r}
setwd(gsea_dir)

prepare_GSEA_rnk_files_direct(deseq_data, gene_sym = '# HGNC', gene_id_col = "HuSym")
# ipa_data <- format_IPA_data(stat_data, gene_col = 'HGNC symbol')
# write.csv(ipa_data, file = 'ipa_data.csv')
```

```{r}
setwd(gsea_dir)

rnk_files <- list.files() %>% grep('.rnk', ., value = TRUE)
comparisons <- rnk_files %>% gsub(pattern = '.rnk', replacement = '')
collections <- c('h', 'c2cp','c5cc','c5mf','c5bp','c2cgp','c8')

gsea_data <- summarize_GSEA_output(directory = 'oct19/', comparisons = comparisons, gene_collections = collections)
write.csv(gsea_data, file = '210909Yil_gsea.csv', row.names = FALSE)

```
